File: Go/ExampleTests/common/cartinput_string.go
package common

type CartInputString struct {
	TotalItems string
	Shipping string
	Discount string
	TotalPrice string
	Notes string
}

func NewCartInputStringFromSlice(v []string) CartInputString {
	s := CartInputString{}
	if len(v) > 0 { s.TotalItems = v[0] }
	if len(v) > 1 { s.Shipping = v[1] }
	if len(v) > 2 { s.Discount = v[2] }
	if len(v) > 3 { s.TotalPrice = v[3] }
	if len(v) > 4 { s.Notes = v[4] }
	return s
}

// NewCartInputStringFromText builds from the text form, e.g. Money as "25 USD".
func NewCartInputStringFromText(text string) CartInputString {
	parts := RequireTokens(text, 5, "CartInput")
	return CartInputString{
		TotalItems: parts[0],
		Shipping: parts[1],
		Discount: parts[2],
		TotalPrice: parts[3],
		Notes: parts[4],
	}
}

func (s CartInputString) String() string {
	return Token(s.TotalItems) + " " + Token(s.Shipping) + " " + Token(s.Discount) + " " + Token(s.TotalPrice) + " " + Token(s.Notes)
}

func (s CartInputString) Equals(o CartInputString) bool {
	return DNCEqual(s.TotalItems, o.TotalItems) &&
		DNCEqual(s.Shipping, o.Shipping) &&
		DNCEqual(s.Discount, o.Discount) &&
		DNCEqual(s.TotalPrice, o.TotalPrice) &&
		DNCEqual(s.Notes, o.Notes)
}

func EqualCartInputStringSlices(a, b []CartInputString) bool {
	if len(a) != len(b) { return false }
	for i := range a {
		if !a[i].Equals(b[i]) { return false }
	}
	return true
}

File: Rust/ExampleTests/common/cartinput_string.rs
#![allow(dead_code, unused_imports, unused_variables)]

#[derive(Debug, Clone, Default)]
pub struct CartInputString {
    pub totalitems: String,
    pub shipping: String,
    pub discount: String,
    pub total_price: String,
    pub notes: String,
}

impl CartInputString {
    pub fn from_vec(v: &[&str]) -> Self {
        Self {
            totalitems: v.get(0).copied().unwrap_or("").to_string(),
            shipping: v.get(1).copied().unwrap_or("").to_string(),
            discount: v.get(2).copied().unwrap_or("").to_string(),
            total_price: v.get(3).copied().unwrap_or("").to_string(),
            notes: v.get(4).copied().unwrap_or("").to_string(),
        }
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    pub fn from_text(text: &str) -> Self {
        let parts = crate::common::tokens::require(text, 5, "CartInput");
        Self {
            totalitems: parts[0].clone(),
            shipping: parts[1].clone(),
            discount: parts[2].clone(),
            total_price: parts[3].clone(),
            notes: parts[4].clone(),
        }
    }
}

impl std::fmt::Display for CartInputString {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", [crate::common::tokens::token(&self.totalitems), crate::common::tokens::token(&self.shipping), crate::common::tokens::token(&self.discount), crate::common::tokens::token(&self.total_price), crate::common::tokens::token(&self.notes)].join(" "))
    }
}

impl PartialEq for CartInputString {
    fn eq(&self, other: &Self) -> bool {
        crate::common::dnc_equal(&self.totalitems, &other.totalitems)
            && crate::common::dnc_equal(&self.shipping, &other.shipping)
            && crate::common::dnc_equal(&self.discount, &other.discount)
            && crate::common::dnc_equal(&self.total_price, &other.total_price)
            && crate::common::dnc_equal(&self.notes, &other.notes)
    }
}
